### ISSUE TITLE
order of imports: bug or wanted behaviour

### ISSUE THREAD

**Issue by juliusstoerrle** [ngxs/store#375]
Today I encountered a problem with the `ngxsOnInit()` life-cycle event not being called in some modules.
After some time and I noticed I imported my feature modules before doing the import of `NgxsStoreModule.forRoot()` in my `app.module.ts` file. Changing the order of imports resolved my problems.

@yarrgh helped me over at Slack and he proposed we start some discussion here to get your options on whether or not the order of imports matters?

### Observed behavior

If you import a FeatureModule before `NgxsStoreModule.forRoot()` in your `app.module`, the life-cycle event `ngxOninit() `will not be called.

### Desired behavior

The RouterModule allows you to call `forChild()` before `forRoot()` so someone could expect the same from NGXS.

### Docs

If we don't change the behavior or in the meantime I suggest we add a note in the docs below:
> It's important that you add NgxsModule.forRoot([]) at the root of your module even if all of your states are feature states.
-(https://ngxs.gitbook.io/ngxs/getting-started/installation)

Maybe also on the [Lazy Loading](https://ngxs.gitbook.io/ngxs/advanced/lazy-loading) Page even if its not directly related to that.
If you like I can create a PR for this?

**Link to thread on Slack**: https://ngxs.slack.com/archives/CA0NS79MG/p1525977069000549

**Comment by amcdnl** [comment]
Interesting. TBH though, why are you calling forFeature in that scenario though? Seems like those should be in the root if they aren't lazy loaded.

**Comment by juliusstoerrle** [comment]
Because I might use that module in multiple apps and it is easier to maintain, develop, understand the setup when I only include the module in the imports and not have to also add the state class in the core/app module. Also there might be (haven't got that case yet) a situation where you don't want to expose your state class to the outside of the library?

But the main point here is I think the initial developer experience, as its not immediately obvious that you imported in the wrong order.

**Comment by StevenLiekens** [comment]
I can confirm that `ngxsOnInit` is not called in `forFeature` state classes if the feature module is not lazy loaded and also imported before importing `NgxsModule.forRoot`.

```ts
@State({
    name: 'feature',
    defaults: {
    }
})
export class FeatureState implements NgxsOnInit {
    ngxsOnInit(ctx) {
        // never reached
    }
}

@NgModule({
  imports: [
    NgxsModule.forFeature([FeatureState])
 ]
})
export class FeatureModule{}

@NgModule({
  imports: [
    FeatureModule,
    NgxsModule.forRoot([])
 ]
})
export class AppModule {}
```

My reason for using `forFeature` without lazy loading is that I can't use lazy loading in my app because of poor design decisions. I still want to use forFeature to decouple feature states from the root module.

I intend to fix the issues and then enable lazy loading at which point this will become a non-issue for me. But right now this is pretty annoying.

**Comment by juliusstoerrle** [comment]
I think a few more people stumbled over this in the past weeks. @amcdnl, @markwhitfeld do you want to fix this (I don't know how some would do that) or just note this in the docs?

**Comment by markwhitfeld** [comment]
@juliusstoerrle I have an idea of how it could be fixed although that would mean a bit of a rewire of the initialization process. This will be fixed, but is just one of the items on y list at the moment.
Could you do a PR for the note if the docs for us?

**Comment by splincode** [comment]
reproduce: https://stackblitz.com/edit/ngxs-simple-xjaodj

**Comment by splincode** [comment]
Released in 3.5.0
